### ISSUE TITLE
Make it possible for opencensus to pass all remaining data to registered exporters when program exits.

### ISSUE THREAD

**Issue by lychung83** [census-instrumentation/opencensus-go#862]
Problem:

When exiting a program running opencensus monitoring, some of the data are not passed to exporter. This happens because as one can see in following code, reportUsage() is called only by timer, so when program exists, all data passed to opencensus after last timer firing event are simply lost.

```
func (w *worker) start() {
	for {
		select {
		case cmd := <-w.c:
			cmd.handleCommand(w)
		case <-w.timer.C:
			w.reportUsage(time.Now())
		case <-w.quit:
			w.timer.Stop()
			close(w.c)
			w.done <- true
			return
		}
	}
}
```

Solution proposal:
1. Make `func (w *worker) stop()` to be (indirectly) callable by client of package.
2. Make `func (w *worker) start()` to call `w.reportUsage()` before returning when `<-w.quit` is triggered.

**Comment by odeke-em** [comment]
Thank you for filing this issue @lychung83!

So currently, `(*worker).stop()` is NOT called by any function nor is `case <-w.quit` triggered by any exit condition. The app just exits if exiting.

I think the idiomatic fix here would to add a global function to view called say `End()` which will invoke `(*worker)(defaultWorker).stop` and that should involve passing `w.reportUsage` as you've suggested. However, that's going to cause a send on closed channel panic with the current stateful design of the worker.

/cc @rakyll @ramonza

**Comment by rakyll** [comment]
I think we need a global Flush function that reports all the collected points.

**Comment by odeke-em** [comment]
@rakyll perhaps I am bikeshedding or not, on the name: I thought of `Flush` too, but that name implies that continued usage after its invocation is allowed, yet we want a function that is terminally invoked, hence the recommendation of `End` :)

**Comment by rakyll** [comment]
Flush function I was suggesting is not stopping the worker, it only flushes the unprocessed data points. Exporting a function that stops the worker might need more thinking because we never planned to give control of the worker to the user entirely.

Also, in practice, should you care about what's collected after you flushed due to the termination signal?

**Comment by odeke-em** [comment]
> Flush function I was suggesting is not stopping the worker, it only flushes the unprocessed data points. 

Roger that, that would work. And in deed, users can invoke `defer view.Flush()` before their main exits so it kills 2 birds with 1 stone.

> Also, in practice, should you care about what's collected after you flushed due to the termination signal?

@lychung83 is this an actual problem that you've encountered or were you hypothesizing based off reading the code? I too don't think it should matter, and moreoever we haven't even provided a stop method nor does the worker know when we are terminating except by the Go runtime cleaning up.

**Comment by lychung83** [comment]
To me, Flush() (not stopping the worker) seems enough. My use cases are:
1. used at the end of the program to report any remaining data points in the worker.
2. used for the test program to make sure all data points put by any measure.M() call to the exporter inspected by test w/o calling time.Sleep() for certain amount of time that's enough to fire worker.c

**Comment by odeke-em** [comment]
Okay, am retitling this issue then as Flush is request that is different from the original request of when a program is exiting.

**Comment by razvanh** [comment]
This will solve https://github.com/census-instrumentation/opencensus-go/issues/862 too.

**Comment by ramonza** [comment]
I don't think we should have a separate function to flush each Exporter and one to flush the internal stats buffer. We should have a single global Flush/Shutdown/Close function that only returns when everything has been flushed. It's also important that things get flushed in the right order: flushing the internal stats before flushing the buffer in the exporter will not work as intended. Users should not have to find this out themselves.

**Comment by rghetia** [comment]
For metrics, concept of Reader was added recently. With reader one can call [ReadAndExport](https://github.com/census-instrumentation/opencensus-go/blob/5ae9166f90ff989c557e45fcb9592ac59afa0fa0/metric/metricexport/reader.go#L177) to export all data on program termination. It would work same as flush.

**Comment by marwan-at-work** [comment]
Hi @odeke-em,  I noticed one of the commits mentioned in this issue addresses this issue directly, just wondering if you'll eventually open it as a PR or if there's anything blocking it. Thank you! 🎉

**Comment by odeke-em** [comment]
Hey Marwan,
   Thank you for the ping! I did open up a PR but it got blocked because
there was no consensus/movement to add it via OpenCensus Specs and I
believe these libraries are now on auto pilot mode and no longer adding
features so unfortunately I don’t think it’ll be done. My PR was sent for
over a year.

**Comment by marwan-at-work** [comment]
@odeke-em Thank you for the response. This seems to me more of a bug than a feature since OpenCensus swallows metrics when a process exists. 

In the same spirit, does that mean all other features and bug requests are frozen? 

I recently opened this issue which, if my assumption is correct, would be quite noteworthy to address: https://github.com/census-instrumentation/opencensus-go/issues/1181

Thanks again :)
